### PR TITLE
Add extent and aspect to images

### LINKED ISSUE #458: Add extent and change they way aspect is calculated

### LINKED ISSUE #540: Fix dose display
Fix dose display

_Originally posted by @grzanka in https://github.com/pytrip/pytripgui/pull/512#pullrequestreview-760071187_

### PR DESCRIPTION
**Pull request by rozchlastywacz**
Retrived code from old branch, but it does not work properly, as contours are plotted outside the image, further investigation is needed

closes #458 
closes #540 
- #458 
- #540

### CODE REVIEW

**Review by grzanka — CHANGES_REQUESTED** [review]
fix failing tests (pep8/flake issues)

File: pytripgui/canvas_vc/objects/ctx.py [hunk 1]
```diff
@@ -22,6 +22,8 @@ def prepare_data_to_plot(self):
 
     def _set_aspect(self):
         if self.projection_selector.plane == "Transversal":
-            self.aspect = 1.0
-        else:
-            self.aspect = self.cube.slice_distance / self.cube.pixel_size
+            self.aspect = self.cube.dimx/self.cube.dimy
```
Comment (grzanka) [inline pytripgui/canvas_vc/objects/ctx.py:25]: The code looks the same in all classes, could it be moved to a mother class ?

File: pytripgui/canvas_vc/objects/ctx.py [hunk 2]
```diff
@@ -22,6 +22,8 @@ def prepare_data_to_plot(self):
...
+        elif self.projection_selector.plane == "Sagittal":
```
Comment (grzanka) [inline pytripgui/canvas_vc/objects/ctx.py:26]: Please add more comments explaining:
  - what is aspect in general, is it `vertical / horizontal` or `horizontal / vertical` ?
For each of the views (transversal, sagittal, coronal) write down explicitly in a comment message what is on vertical axis and what on horizontal (X, Y or Z ?).

This information could be also added as the code (assuming some convention):
```
if  transveral:
   vertical_size = dimx * pixel_size
   horizontal_size = dimy * pixel size
(...)
aspect = vertical_size / horizontal_size
```

File: pytripgui/canvas_vc/plotter/images/patient_image_base.py [hunk 3]
```diff
@@ -53,3 +53,22 @@ def is_present(self) -> bool:
...
+
+    @staticmethod
+    def calculate_extent(data):
+        """
+        Returns extent for passed data.
+        """
+        # get minimal X, Y and Z coordinates
+        min_x_mm, min_y_mm, min_z_mm = data.cube.indices_to_pos([0, 0, 0])
+        # get maximal X, Y and Z coordinates
+        max_x_mm, max_y_mm, max_z_mm = data.cube.indices_to_pos([data.cube.dimx, data.cube.dimy, data.cube.dimz - 1])
```
Comment (grzanka) [inline pytripgui/canvas_vc/plotter/images/patient_image_base.py:65]: Explain why `-1` is needed for Z

**Review by grzanka — CHANGES_REQUESTED** [review]

File: pytripgui/canvas_vc/objects/ctx.py (hunk 1 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/objects/ctx.py:25]: Superclass does not exist, but I can make one as a little refactor

**Comment by rozchlastywacz** [comment]
image
Left axis is not as pretty as bottom one plotted, I think it's because there is not enough space, but it will be fixed in #429

**Comment by rozchlastywacz** [comment]
Also DeepSource keeps failing, but something is wrong with it, and it does not say whats wrong:
image
After clicking *here*
image

File: pytripgui/canvas_vc/plotter/images/patient_image_base.py (hunk 3 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/plotter/images/patient_image_base.py]: Done

File: pytripgui/canvas_vc/objects/ctx.py (hunk 2 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/objects/ctx.py]: Done

File: pytripgui/canvas_vc/objects/ctx.py (hunk 1 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/objects/ctx.py]: Done

File: pytripgui/canvas_vc/objects/data_base.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+from abc import ABC, abstractmethod
+from typing import Optional
+
+from pytrip import Cube
+
+
+class PlotDataBase(ABC):
+    def __init__(self, selector):
+        self.aspect = 1.0  # aspect ratio of plot
```
Comment (grzanka) [inline pytripgui/canvas_vc/objects/data_base.py:9]: Add a type hint here

File: pytripgui/canvas_vc/plotter/images/ctx_image.py [hunk 5]
```diff
@@ -11,9 +11,28 @@ def __init__(self, axes):
...
                                         cmap=self._colormap,
                                         vmin=data.contrast_ct[0],
                                         vmax=data.contrast_ct[1],
                                         aspect=data.aspect,
+                                        extent=extent,
+                                        origin='lower',
                                         zorder=self.zorder)
+        self._set_labels(data)
+
+    def _set_labels(self, data: Ctx):
```
Comment (grzanka) [inline pytripgui/canvas_vc/plotter/images/ctx_image.py:25]: Add a comment explaining why labels are set only on Ctx and not on a Dos and LET (cause the other simulation results are just overlays, right?).

File: pytripgui/canvas_vc/plotter/managers/voi_manager.py [hunk 6]
```diff
@@ -63,24 +62,15 @@ def plot_voi(self, vdx: Vdx):
...
-                #   but i think it fits better if you look on plotted contours
-                z_offset = min(vdx.ctx.slice_pos) - vdx.ctx.slice_distance  # slice_pos contains z positions in mm
-
-                # contours are in [[x0,y0,z0], [x1,y1,z1], ... [xn,yn,zn]] (mm), we need to remove offsets
-                data = np.array(_c.contour) - np.array([vdx.ctx.xoffset, vdx.ctx.yoffset, z_offset])
-
-                # translating positions in mm into pixel positions
-                data_pixels = self._plane_points_idx(data, vdx.ctx, vdx.projection_selector.plane)
+                # getting data to plot directly, without any conversions, subtracting offsets and so on
+                # stored in terms of millimeters
```
Comment (grzanka) [inline pytripgui/canvas_vc/plotter/managers/voi_manager.py:66]:
```suggestion
                # stored in units of millimeters

```

**Review by grzanka — CHANGES_REQUESTED** [review]
minor changes requested

File: pytripgui/canvas_vc/plotter/managers/voi_manager.py (hunk 6 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/plotter/managers/voi_manager.py]: Done

File: pytripgui/canvas_vc/plotter/images/ctx_image.py (hunk 5 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/plotter/images/ctx_image.py]: Done by changing the place where method is defined and invoked

File: pytripgui/canvas_vc/objects/data_base.py (hunk 4 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/objects/data_base.py:11]: Done

**Comment by grzanka** [comment]
Dose display seems to be completely broken:
image

**Comment by grzanka** [comment]
image

**Review by grzanka — CHANGES_REQUESTED** [review]
Fix dose display

**Comment by grzanka** [comment]
@rozchlastywacz I think it is right time now to resume work on this one

File: pytripgui/canvas_vc/plotter/images/patient_image_base.py [hunk 7]
```diff
@@ -53,3 +60,28 @@ def is_present(self) -> bool:
...
+        min_x_mm, min_y_mm, min_z_mm = cube.indices_to_pos([0, 0, 0])
+        # get maximal X, Y and Z coordinates
+        # 'data.cube.dimz - 1' is described in https://github.com/pytrip/pytrip/issues/592
+        max_x_mm, max_y_mm, max_z_mm = cube.indices_to_pos([cube.dimx, cube.dimy, cube.dimz - 1])
+        plane = data.projection_selector.plane
+        # depending on plane, return proper list those above
+        # extent = [horizontal_min, horizontal_max, vertical_min, vertical_max]
+        # "Transversal" (xy)
+        if plane == "Transversal":
+            return [min_x_mm, max_x_mm, min_y_mm, max_y_mm]
```
Comment (grzanka) [inline pytripgui/canvas_vc/plotter/images/patient_image_base.py:79]:
```suggestion
            return (min_x_mm, max_x_mm, min_y_mm, max_y_mm)

```

I'd reccomend immutable tuples as return type

**Review by grzanka — APPROVED** [review]
Just a tiny comment, apart from that OK

**Review by grzanka — APPROVED** [review]
LGTM

**Pull request merged**